packages/web-frontend/src/sagas.js

```diff
@@ -117,9 +130,124 @@ import {
   selectOrgUnit,
   selectOrgUnitChildren,
   selectOrgUnitCountry,
+  selectProjectByCode,
 } from './selectors';
 import { formatDateForApi, isMobile, processMeasureInfo } from './utils';
 import { getDefaultDates } from './utils/periodGranularities';
+import { fetchProjectData } from './projects/sagas';
+import { clearLocation } from './historyNavigation/historyNavigation';
+import { decodeLocation } from './historyNavigation/utils';
+import { PASSWORD_RESET_PREFIX, VERIFY_EMAIL_PREFIX } from './historyNavigation/constants';
+
+function* watchFetchInitialData() {
+  yield take(FETCH_INITIAL_DATA);
+
+  // Login must happen first so that projects return the correct access flags
+  yield call(findUserLoggedIn, LOGIN_TYPES.AUTO);
+  yield call(fetchProjectData);
+  yield call(handleLocationChange, {
+    location: getInitialLocation(),
+    previousLocation: clearLocation(),
+  });
+}
+
+function* handleInvalidPermission({ projectCode }) {
+  const state = yield select();
+  const { isUserLoggedIn } = state.authentication;
+
+  if (isUserLoggedIn) {
+    // show project access dialog
+    const project = selectProjectByCode(state, projectCode);
+
+    if ('code' in project) {
+      yield put(setRequestingAccess(project));
+      yield put(setOverlayComponent('requestProjectAccess'));
+      return;
+    }
+
+    // handle 404s
+    console.log('project does not exist - 404'); // Todo: handle 404s
+    return;
+  }
+  // show login dialog
+  yield put(setOverlayComponent(LANDING));
+}
+
+function* setComponentsIfUpdated(components, previousComponents, otherComponents) {
+  for (let i = 0; i < components.length; i++) {
```

Comment:
I guess you can also use a `forEach` here

Reply:
Unfortunately you can not call yield from within an inner function. It must be from within the main generator function.

Reply:
No worries!